kiosk/examples/unlock.move

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[test_only]
+/// This module contains a simple `unlocker` module which enables creators to
+/// `unlock` (as opposed to `lock`) assets without fulfilling the default set
+/// of requirements (rules / policies).
+module kiosk::unlock_example {
+    use sui::kiosk::{Self, Kiosk, KioskOwnerCap};
+    use sui::package::{Self, Publisher};
+    use sui::object::{Self, ID, UID};
+    use sui::tx_context::TxContext;
+    use sui::transfer_policy::{
+        Self as policy,
+        TransferPolicy,
+        TransferPolicyCap
+    };
+    use sui::coin;
+
+    /// Trying to destroy the Unlocket object while not being its publisher.
+    const ENotPublisher: u64 = 0;
+
+    /// An unlocker is a special type of object which can be used to unlock
+    /// assets with another `TransferPolicy`.
+    struct Unlocker<phantom T> has key, store {
+        id: UID,
+        policy: TransferPolicy<T>,
+        cap: TransferPolicyCap<T>,
+    }
+
+    /// Create a new `Unlocker` object; can either be used directly or wrapped
+    /// into a custom object (with matching logic).
+    public fun new<T: key + store>(
+        publisher: &Publisher,
+        ctx: &mut TxContext
+    ): Unlocker<T> {
+        let (policy, cap) = policy::new(publisher, ctx);
```

Comment:
Checking my understanding: the check that `T` is from `publisher`'s package happens in `policy::new` because by type inference, we know that `policy` and `cap` must be parameterised by the same `T`?

Reply:
Exactly! That is how it happens.